libmamba/src/core/package_handling.cpp

```diff
@@ -142,24 +185,44 @@ namespace mamba
         }
         fs::current_path(directory);
 
-        for (auto& dir_entry : fs::recursive_directory_iterator("."))
+        std::vector<std::pair<fs::u8path, std::pair<int, std::string>>> files;
+        if (ca != compression_algorithm::zip)
         {
-            // we only add empty directories to the archive
-            if (dir_entry.is_directory() && !fs::is_empty(dir_entry.path()))
+            for (auto& dir_entry : fs::recursive_directory_iterator("."))
             {
-                continue;
+                auto clean_path = dir_entry.path().lexically_relative("./");
+                files.push_back({ clean_path, order(clean_path) });
             }
-
-            std::string p = dir_entry.path().string();
-            // do this in a better way?
-            if (p[0] == '.')
+        }
+        else
+        {
+            // for zip files, sort `info` last
+            for (auto& dir_entry : fs::directory_iterator("."))
             {
-                p = p.substr(1);
+                auto clean_path = dir_entry.path().lexically_relative("./");
+                files.push_back({ clean_path, zip_order(clean_path) });
             }
-            if (p[0] == '/')
+        }
+
+        std::sort(files.begin(),
+                  files.end(),
+                  [](const auto& a, const auto& b) { return a.second < b.second; });
```

Comment:
replacing this lambda with:
```cpp
[](const auto& a, const auto& b) { return a.second < b.second || (a.second == b.second && a.first < b.first); }
```
allows to return an `int` instead of `std::pair<int, std::string>` in the `order` and  `zip_order` function, and simplifies the files type.

Reply:
or even simpler:

```cpp
[](const auto& a, const auto& b) { return std::tie(a.second, a.first) < std::tie(b.second, b.first); }
```

Reply:
Is the comparison object even necessary? `std::pair` already have lexicographic comparison using both members by default.

Reply:
yes, exactly. I think what's confusing is that this is a pair of pairs and the latter pair is actually used with it's default comparison operator.

Reply:
Ah ok, I think I see what you want now :) will change it.

Reply:
lambda is gone :)

Reply:
If the order functions return `int` instead of `std::pair<int, string>` then we want to sort by the second member of the pair (the int)  first, and then the first (the path). Therefore the lambda is still required since the natural comparison of `std::pair` starts with the first member and then the second if the first ones are equal.